app/services/fact_importer.rb

```diff
@@ -190,11 +190,16 @@ def normalize(facts)
 
   def db_facts
     query = host.fact_values
+    # MySQL query optimizer does not appear to pick the correct index here: https://projects.theforeman.org/issues/25053
     if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'mysql'
-      # MySQL query optimizer does not appear to pick the correct index here: https://projects.theforeman.org/issues/25053
       query = query.from("fact_values USE INDEX(index_fact_values_on_fact_name_id_and_host_id)")
     end
-    query.where(:fact_name => fact_names.values).reorder('')
+    # When using PostgreSQL use regular expression to prevent query with thousands of parameters
+    if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'postgresql'
+      query.where("fact_name ~ ?", "^(#{fact_names.values.join('|')})$").reorder('')
```

Comment:
Won't we need any escaping on the fact names in case they have regex chars in them?

Reply:
Good point, I think it should be pretty sane approach to use just Ruby's `Regexp.escape` chances for edge cases are pretty low.